Feature 2419 use shared expansion panel layer categories view

ianguerin opened the pull request:
This is a clean up tasks from implementing expansion panels for zoom-to presets feature

yvonnesjy commented on src/js/views/maps/LayerCategoryListView.js:
```diff
@@ -47,18 +53,30 @@ define(
           if (options.collection instanceof AssetCategories) {
-            this.layerCategoryItemViews = options.collection.map(categoryModel => {
-              return new LayerCategoryItemView({model: categoryModel});
+            this.panelsModel = new ExpansionPanelsModel({ isMulti: true });
```
does this need to be an instance variable?

ianguerin replied:
It does not, making local

yvonnesjy commented on src/js/templates/maps/expansion-panel.html:
```diff
@@ -1,6 +1,8 @@
-    <i class="icon <%= icon %>"></i>
+    <%if(!isSvgIcon) {%>
+      <i class="icon <%= icon %>"></i>
+    <%}%>
```
does it work if we just do

```
<% if (isSvgIcon) { %>
  <%= icon %>
<% } else { %>
  <i ...></i>
<% } %>
```

ianguerin replied:
Combined this with default SVG icon

yvonnesjy replied:
ah the default is a layer icon that might only make sense for layer categories. I'm not sure if we want it to apply to all expansion panel. might be better if we pass it in from layer category list?

ianguerin replied:
Solved by resolving a TODO in AssetCategory.js setting default icon

yvonnesjy commented on src/js/views/maps/ExpansionPanelView.js:
```diff
@@ -70,16 +71,30 @@ define(
           this.contentViewInstance = contentViewInstance;
           this.panelsModel = panelsModel;
           this.startOpen = !!startOpen;
+          this.icon = icon;
```
can we just reference this.templateVars.icon instead of creating another variable?

ianguerin replied:
Yep, done

yvonnesjy commented on src/js/views/maps/LayerCategoryListView.js:
```diff
@@ -82,8 +100,21 @@ define(
+            } else {
+              panel.$el.hide();
+              panel.collapse();
+            }
```
Can this behavior be moved into expansion panel? Is it generalizable?

ianguerin replied:
I don't think so, expansion panels in general shouldn't have anything to do with search. These do, because the components inside of them (LayerListView) are considered in layer search

yvonnesjy commented on src/js/templates/maps/layer-category-item.html:
```diff
@@ -1,13 +1 @@
-<div class="<%= classNames.metadata %>">
-  <span class="<%= classNames.icon %>">
-    <svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 24 24"><path d="m3.2 7.3 8.6 4.6a.5.5 0 0 0 .4 0l8.6-4.6a.4.4 0 0 0 0-.8L12.1 3a.5.5 0 0 0-.4 0L3.3 6.5a.4.4 0 0 0 0 .8Z"/><path d="M20.7 10.7 19 9.9l-6.7 3.6a.5.5 0 0 1-.4 0L5 9.9l-1.8.8a.5.5 0 0 0 0 .8l8.5 5a.5.5 0 0 0 .5 0l8.5-5a.5.5 0 0 0 0-.8Z"/><path d="m20.7 15.1-1.5-.7-7 3.8a.5.5 0 0 1-.4 0l-7-3.8-1.5.7a.5.5 0 0  ...
```
we should preserve the default value for layer category

ianguerin replied:
I guess we missed a unit test. Combining this with your comment above about simplifying the template and the way icons are rendered. I'll svg icon if isSvgIcon otherwise icon as a font awesome icon if it is defined, otherwise, render the default SVG

robyngit commented on src/js/views/maps/ExpansionPanelView.js:
```diff
@@ -70,16 +71,30 @@ define(
-        initialize({ title, contentViewInstance, icon, panelsModel, startOpen }) {
-          this.templateVars.title = title;
-          this.templateVars.icon = icon;
+        initialize({ title, contentViewInstance, icon, panelsModel, startOpen, isSvgIcon, }) {
```
```suggestion
        /**
         * Initialize the view with the given options.
         * @param {ExpansionPanelViewOptions} options The options for this view.
         */
        initialize({ title, contentViewInstance, icon, panelsModel, startOpen, isSvgIcon, }) {
```

robyngit commented on src/js/views/maps/ExpansionPanelView.js:
```diff
@@ -70,16 +71,30 @@ define(
-          this.templateVars.icon = icon;
+        initialize({ title, contentViewInstance, icon, panelsModel, startOpen, isSvgIcon, }) {
+          this.templateVars = {
+            ...this.templateVars,
```
```suggestion
            classNames: CLASS_NAMES,
```

and then you could delete lines 53-58:
```js
        /** Values meant to be used by the rendered HTML template. */
        templateVars: {
          classNames: CLASS_NAMES,
          title: '',
          icon: '',
        },
```

ianguerin commented on src/js/views/maps/ExpansionPanelView.js:
```diff
@@ -24,8 +24,8 @@ define(
      * @extends Backbone.View
-     * @screenshot views/maps/viewfinder/ExpansionPanelView_closed.png
-     * @screenshot views/maps/viewfinder/ExpansionPanelView_open.png
+     * @screenshot views/maps/ExpansionPanelView_closed.png
+     * @screenshot views/maps/ExpansionPanelView_open.png
```
I messed this up somehow, I only have a single screenshot, but it's not referenced here! Fixing

yvonnesjy commented on src/js/views/maps/ExpansionPanelView.js:
```diff
@@ -134,7 +130,7 @@ define(
           if (this.templateVars.isSvgIcon) {
-            this.insertIcon(this.icon);
+            this.insertIcon(this.templateVars.icon);
           }
```
this is covered in the template. no need to insert in js anymore

ianguerin replied:
Done

ianguerin commented:
Alright, this PR should  be good to give a final review through, rebased onto develop and fixed formatting issues. I *think* the lint issues left are skippable?

robyngit commented on src/js/views/maps/LayerCategoryListView.js:
```diff
@@ -42,36 +48,41 @@ define([
+       * @type {ExpansionPanel[]}
        */
-      layerCategoryItemViews: undefined,
+      panels: undefined,
```
maybe:

```suggestion
      panels: [],
```

to prevent errors (e.g. with `_.reduce`) if `options.collection` isn't passed to the view and then `view.panels` is undefined

ianguerin replied:
Done

robyngit commented on src/js/views/maps/LayerCategoryListView.js:
```diff
@@ -42,36 +48,41 @@ define([
-              return new LayerCategoryItemView({ model: categoryModel });
-            },
-          );
+          this.panels = options.collection.map((categoryModel) => {
+            const icon = categoryModel.get("icon");
+            return new ExpansionPanelView({
+              contentViewInstance: new LayerListView({
+                collection: categoryModel.get("mapAssets"),
+                isCategorized: true,
+              }),
+              icon,
+              isSvgIcon: IconUtilities.isSVG(icon),
+              panelsModel: new ExpansionPanelsModel({ isMulti: true }),
+              title: categoryModel.get("label"),
+            });
+          });
```
could we just save `options.collection` to the view then instantiate the `ExpansionPanelView`s in `render`? I ask because it would be a more typical pattern (unusual to instantiate sub views in `initialize`) and also avoids iterating through the collection/panels twice.

ianguerin replied:
Done

robyngit commented on src/js/views/maps/LayerCategoryListView.js:
```diff
@@ -42,36 +48,41 @@ define([
       /**
        * Executed when a new LayerCategoryListView is created
-       * @param {Object} options - A literal object with options to pass to the view
+       * @param {object} options - A literal object with options to pass to the view
```
```suggestion
       * @param {object} options - A literal object with options to pass to the view
       * @param {AssetCategories} options.collection - The collection of AssetCategories to display
```

ianguerin replied:
Done

robyngit commented on src/js/views/maps/LayerCategoryListView.js:
```diff
@@ -5,16 +5,22 @@ define([
+  "views/maps/ExpansionPanelView",
+  "models/maps/ExpansionPanelsModel",
+], (
   $,
```
could remove jquery as a dependency

ianguerin replied:
Done

robyngit reviewed (approved):
I tested this out and it works great! No issues. Added little optional suggestions only.

ianguerin commented:
@yvonnesjy, please take another look - can't merge as it says you have requested changes

yvonnesjy commented on src/js/templates/maps/layer-category-item.html:
This template can be deleted now?

ianguerin replied:
Done
